<!DOCTYPE HTML PUBLIC "-//WAPFORUM//DTD WML 1.2//EN">
<html><head><title>BRaliBase III</title>
  
  <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
  <meta name="GENERATOR" content="Quanta Plus"></head>
<body>
<hr>
<h1>BRaliBase III.</h1>
<hr>


<p>Freyhult EK, Bollback JP and Gardner PP (2006) Exploring genomic dark
matter: homology search for non-coding RNA. <i>In preparation.</i> <br> <a
href="FreyhultBollbackGardner06.pdf">Preprint</a> [UPDATED: 23/05/06].
<br> <a href="Supplementary_Materials.pdf">Supplementary materials.</a> </p>

<hr>

<h2>Supplementary Information.</h2>

<p>A <a href="../bralibase3.tar.gz">tarball</a> of the data-sets.</p>


<hr>

<p>The 5 and 20 sequence query and database datasets can be found <a href="datasets/">here.</a></p>
<p>The <a href="scripts/">scripts</a> used to run and analyse homology search software.</p>
<p>The tab-delimited <a href="results/">results</a> files from aforementioned homology search software.</p>
<p>The <a href="README">README</a> giving details of the datasets and
summarising scripts and results.</p>

<hr>
<hr>


<P><h4>The following algorithms were compared in this study:</h4></P>
<hr>

<center>
<table title="Plan 0" align="left" columns="2" BORDER=1>
	<COL WIDTH=55*>
	<COL WIDTH=201*>


<tr>
<td><P><b><a name="singleSeq">Single sequence methods</a></b><A HREF=""></A></P></td>
<td></td>
</tr>


<tr>
<td><P><A HREF="http://www.ncbi.nlm.nih.gov/BLAST/">NCBI-BLAST</A></P></td>
<td>finds regions of local similarity between sequences. The program compares nucleotide or protein sequences to sequence databases and calculates the statistical significance of matches.</td>
</tr>

<tr>
<td><P><A HREF="http://blast.wustl.edu">WU-BLAST</A></P></td>
<td>Faster at any sensitivity, more sensitive at any speed, the original gapped BLAST with statistics, providing the performance, features and reliability demanded by technical professionals.</td>
</tr>

<tr>
<td><P><A HREF="http://fasta.bioch.virginia.edu/">FASTA</A></P></td>
<td>Compares a protein sequence to another protein sequence or to a protein database, or a DNA sequence to another DNA sequence or a DNA library.</td>
</tr>

<tr>
<td><P><A HREF="http://www.paralign.org/">PARALIGN</A></P></td>
<td>Rapid and sensitive sequence similarity searches powered by parallel computing technology.
</td>
</tr>

<tr>
<td><P><A HREF="http://fasta.bioch.virginia.edu/">SSEARCH</A></P></td>
<td>Performs a rigorous Smith-Waterman alignment between a protein sequence and another protein sequence or a protein database, or with DNA sequence to another DNA sequence or a DNA library (very slow).</td>
</tr>

<tr>
<td><P><b><a name="profileHMM">Profile HMM methods</a></b><A HREF=""></A></P></td>
<td></td>
</tr>

<tr>
<td><P><A HREF="http://hmmer.wustl.edu/">HMMer</A></P></td>
<td>Profile hidden Markov models (profile HMMs) can be used to do sensitive database searching using statistical descriptions of a sequence family's consensus. HMMER is a freely distributable implementation of profile HMM software for protein sequence analysis.</td>
</tr>

<tr>
<td><P><A HREF="http://www.cse.ucsc.edu/research/compbio/sam.html">SAM</A></P></td>
<td>A collection of software tools for creating, refining, and using a type of statistical model called a linear hidden Markov model for biological sequence analysis.</td>
</tr>

<tr>
<td><P><b><a name="struct">Structural methods</a></b><A HREF=""></A></P></td>
<td></td>
</tr>

<tr>
<td><P><A HREF="http://tagc.univ-mrs.fr/erpin/">ERPIN</A></P></td>
<td>ERPIN (Easy RNA Profile IdentificatioN) is an RNA motif search program developed by
Daniel Gautheret and Andr&eacute; Lambert
</td>
</tr>

<tr>
<td><P><A HREF="http://www.genetics.wustl.edu/eddy/infernal/">Infernal</A></P></td>
<td>cmalign aligns the RNA sequences in seqfile to the covariance model (CM) in cmfile, and outputs a multiple sequence alignment.</td>
</tr>

<tr>
<td><P><A HREF="http://bio.cs.washington.edu/supplements/zasha-ravenna/">RaveNnA</A></P></td>
<td>A software package for faster covariance models whilst provably sacrificing no accuracy for rigorous filtering. Also includes heuristic filters.</td>
</tr>

<tr>
<td><P><A HREF="http://www.genetics.wustl.edu/eddy/software/#rsearch">RSEARCH</A></P></td>
<td> RSEARCH aligns an RNA query to target sequences, using SCFG algorithms to score both secondary structure and primary sequence alignment simultaneously. It's slow, but somewhat more capable of finding significant remote RNA structure homologies than sequence alignment methods like BLAST.</td>
</tr>


<tr>
<td><P><A HREF="http://exon.umdnj.edu/software/RSmatch/">RSmatch</A></P></td>
<td>RNA Secondary Structure Matcher: provides four functions: (1) regular database search, (2) multiple structure alignment, (3) iterative database search, and (4) pairwise sequence alignment.</td>
</tr>



</table>
</center>



<P><BR>
</P>

<P>.<BR>
</P>


<hr>

    <address>
      <a href="http://www.binf.ku.dk/~pgardner">Paul Gardner</a>, <a href="mailto:ppgardner@bi.ku.dk">&lt;ppgardner@bi.ku.dk&gt;</a><br>
       Dept. of Evolutionary Biology, University of Copenhagen, <br>
       Universitetsparken 15, 2100 Copenhagen Ø, Denmark.
    </address>
    <hr>

    
    
<p class="TIMESTAMP">Last changed: 2006-06-19 13:01:21 pgardner</p>

</body></html>
